### ISSUE TITLE
Problems with first "Hello World"

### ISSUE THREAD

**Issue by basementmedia2** [microsoftgraph/msgraph-sdk-php#759]
Hi all,

I would like to use PHP to access various areas of Outlook 365 (including retrieving emails, retrieving/creating calendar entries, etc).

What i already did:

I installed this SDK via Composer

Then I registered my application via Microsoft Azure Active Directory:
https://portal.azure.com/#blade/Micr...tionsListBlade

There I got my tenantId, ClientID, and the clientSecret.

This is my test-code which does output the accessToken but does not output the user information:

```
require_once DIR . '/vendor/autoload.php';

// Include the Microsoft Graph classes
use Microsoft\Graph\Graph;
use Microsoft\Graph\Model;

// Data from Azure Active Director
$tenantId="xxx";
$clientId="yyy";
$clientSecret="zzz";

$guzzle = new \GuzzleHttp\Client();
$url = 'https://login.microsoftonline.com/' . $tenantId . '/oauth2/token?api-version=1.0';
$token = json_decode($guzzle->post($url, [
'form_params' => [
'client_id' => $clientId,
'client_secret' => $clientSecret,
'resource' => 'https://graph.microsoft.com/',
'grant_type' => 'client_credentials',
],
])->getBody()->getContents());

$accessToken = $token->access_token;

// This works! The Access-Token is echoed
echo "AccessToken:".$accessToken;

// But from here on, i get no output
$graph = new Graph();
$graph->setAccessToken($accessToken);

$user = $graph->createRequest("GET", "/me")
->setReturnType(Model\User::class)
->execute();

print_r($user);

echo "Hello, my name is {$user->getGivenName()}."; 
```

Can you help me to get the code above working?

Furthermore i wonder, if this is the correct way to connect or if i should use api-version 2.0?
`
$url = 'https://login.microsoftonline.com/' . $tenantId . '/oauth2/token?api-version=1.0';
`
If i use this URL, i get no accessToken

`$url = 'https://login.microsoftonline.com/' . $tenantId . '/oauth2/v2.0/token?api-version=2.0';`

Thank you for your support.

Best wishes
Daniel

**Comment by Ndiritu** [comment]
Hi @basementmedia2 

My initial guess is that you may need to set up delegated permissions.

To use the '/me' endpoint you need to have set up delegated permissions on the application you registered.

The alternative would be to call `/users/{yourUserPrincipalName}` in place of '/me'.

Please see the [Get User permissions](https://docs.microsoft.com/en-us/graph/api/user-get?view=graph-rest-1.0&tabs=http#permissions) for more and [Microsoft Graph Permissions](https://docs.microsoft.com/en-us/graph/auth/auth-concepts#microsoft-graph-permissions) for more on setting up the right permissions.

In case you have already set up delegated permissions, then we'd need to see the response payload that you are receiving from the API.

**Comment by basementmedia2** [comment]
Hi,

when i connect to the graph explorer directly (via https://developer.microsoft.com/en-us/graph/graph-explorer) and do this

![grafik](https://user-images.githubusercontent.com/23479553/150074002-56b6a4ad-9386-4c5f-9720-57fde05b7ef5.png)

i get a result.

In Azure the following authorities are set:

![grafik](https://user-images.githubusercontent.com/23479553/150074195-f8c9f615-7ef4-44f2-befe-b671954c2a87.png)

Is that ok for using the example script?

Best wishes
Daniel

**Comment by Ndiritu** [comment]
(Sorry just noticed your question on the auth endpoint), use `"https://login.microsoftonline.com/".$tenantId."/oauth2/v2.0/token"`

Please try executing the request without a return type. In this case you should get back a `GraphResponse` object which contains the status code, headers, body etc. Try to dump that and share the response payload if you get any.

```php
$response= $graph->createRequest("GET", "/me")
                  ->execute();
```

**Comment by basementmedia2** [comment]
Hi

changed the script as follows:
```

require_once __DIR__ . '/vendor/autoload.php';
 
use Microsoft\Graph\Graph;
use Microsoft\Graph\Model;

// Credentials from microsoft
$tenantId="xxx";
$clientId="yyy";
$clientSecret="zz";


$guzzle = new \GuzzleHttp\Client();
//$url = 'https://login.microsoftonline.com/' . $tenantId . '/oauth2/token?api-version=1.0'; // this works
$url = 'https://login.microsoftonline.com/'.$tenantId.'/oauth2/v2.0/token';
$token = json_decode($guzzle->post($url, [
    'form_params' => [
        'client_id' => $clientId,
        'client_secret' => $clientSecret,
        'resource' => 'https://graph.microsoft.com/',
        'grant_type' => 'client_credentials',
    ],
])->getBody()->getContents());

$accessToken = $token->access_token;
echo "AccessToken:".$accessToken; // Does not output access token anymore...

$graph = new Graph();
$graph->setAccessToken($accessToken);

$response= $graph->createRequest("GET", "/me")
                  ->execute();

print_r($response);
```

But now i don't get an accessToken.
Only if i use this endpoint

$url = 'https://login.microsoftonline.com/' . $tenantId . '/oauth2/token?api-version=1.0'

i get one.

also tried this

'scope' = > 'https://graph.microsoft.com/.default',

instead of this

'resource' => 'https://graph.microsoft.com/',

In the Microsoft forum I had received the tip to use this resource when using the Api 2.0.

**Comment by SomeDevWeb** [comment]
Hi. I have the same issue. I try to use this very basic example and I don't get any results. I get a token ok, but I have no idea not ANY WISH to research about all these weird variables like {yourUserPrincipalName} and so on. 
For many people, we are just forced through the job to do some MS integrations that we HATE and we need to spend as little time as possible with this, so I don't understand how MS can't do some step by step simple examples about HOW or WHERE TO GET all this info from!
It's not enough MS destroyed 10 years of my life, now I already lost another 3 days trying to find a way to just create a meeting in outlook and not even the basic example works in the documentation.
Is there a simple explanation of ALL STEPS someone needs to do to even do the basic thing?

**Comment by zengin** [comment]
As mentioned in our README file, https://github.com/microsoftgraph/msgraph-sdk-php#authenticate-with-the-microsoft-graph-service, delegated authentication scenarios are not implemented directly in the PHP SDK. Please refer to this sample for delegated flow: https://github.com/microsoftgraph/php-connect-sample

@basementmedia2, the token acquired using client_secret only holds "application permissions" granted to the application and according to the screenshot above, there is no "application permission" assigned for your application. Check out the corresponding documentation page if there is an "application permission" alternative to access the same data. For example: As @Ndiritu noted, for "Get User" API, one can either use a token with "delegated User.Read permission", or a token with "application User.Read.All permission". `/me` is only allowed in delegated flow, so `/users/<user-id>` path should be used with application permissions.

In Azure Portal, when you click on "Add a permission" you can select the application permissions like below:

![image](https://user-images.githubusercontent.com/803311/151491969-ba55a4f1-185c-43eb-97df-c8c150ab41ee.png)

**Comment by SomeDevWeb** [comment]
Thanks for the info. It's a bit more clear now but it still doesn't work and still, there are quite a few unclarities. 
So, I have enabled the permissions: **User.Read.All Application**. (I have an image but I see I can't upload it here...)
First weird thing is **"Admin consent required" is "Yes"**. Does that mean I need to do something else? There is no other option on the screen to do anything. If yes, who needs to consent. There's just me on this account, so I should be admin, no? Why does that appear there then? What's going on????

Secondly, it says "/users/<user-id>".... So where the HELL is that user-id listed???? And actually, what does it mean???
What kind of ID is it? A token, the user email? What? Where do I get it from?

I tell you, most MS projects are way too convoluted, horrible documentation, all that in the end the platform to be closed and the developers using it to lose years of their life for nothing. MS should pay a fair compensation for all the time we lost or just completely shut down instead of waring everyone's time!

It's unbelievable how an ARMY of tens o thousands of people can't make a SIMPLE VIDEO with freakin few steps for idiots like me that DON'T WANT ANYTHING TO DO WITH MS OR READ 3 MONTH FROM THE MILLIONS OF HYPERTECHNICAL DOCUMENTATION PAGES THAT DON'T MAKE ANY SENSE!!!

Anyway, let's see if we can make any of this work, even a very basic example. 
It's already 5 days I spent on this. I'm very curious when I will be able to create that damn outlook meeting! :(

**Comment by basementmedia2** [comment]
Hi,

you can use the Email-Adress alternatively to the user-id, works  for me.

Best wishes 
Daniel

**Comment by SomeDevWeb** [comment]
**Thanks!** I tried that too and still nothing. 
Now I just got another weird message.

The portal is having issues getting an authentication token. The experience rendered may be degraded.
Additional information from the call to get a token:
Extension: fx
Resource: jarvis
Details: AADSTS50020: User account '{EmailHidden}' from identity provider 'live.com' does not exist in tenant 'Microsoft Services' and cannot access the application 'c44b4....'(Azure Portal) in that tenant. The account needs to be added as an external user in the tenant first. Sign out and sign in again with a different Azure Active Directory user account.
Trace ID: dd08105b...
Correlation ID: deba530b...
Timestamp: 2022-01-28 14:41:49Z

**Comment by SomeDevWeb** [comment]
So, I've asked an admin to grant all permission and I use:
**User.ReadWrite.All | Application** | Read and write all users' full profiles | Yes | Granted for...
**User.Read.All | Application** | Read all users' full profiles | Yes |  Granted for...

What other permission do I need please or what am I doing wrong?

I always get a valid token, but always a blank screen after, no other response, not even an error.

Here is my code: 

use Microsoft\Graph\Graph;
use Beta\Microsoft\Graph\Model as BetaModel; 

$graph = new Graph();

$graph->setAccessToken($ms_accessToken);

$user = $graph->setApiVersion("beta")
  ->createRequest("GET", "/users/[my_user_email]")
  ->setReturnType(BetaModel\User::class)
  ->execute();

echo "Hello, I am $user->getGivenName() ";


Thank you!

**Comment by basementmedia2** [comment]
Hi,

try this:

```
// Include the Microsoft Graph classes
use Microsoft\Graph\Graph;
use Microsoft\Graph\Model;
$tenantId="GET THIS IN AZURE USER AREA (YOUR REGISTERED APP)";
$clientId="GET THIS IN AZURE USER AREA (YOUR REGISTERED APP)";
$clientSecret="GET THIS IN AZURE USER AREA (YOUR REGISTERED APP)";
$guzzle = new \GuzzleHttp\Client();
$url = 'https://login.microsoftonline.com/' . $tenantId . '/oauth2/token?api-version=1.0';
$token = json_decode($guzzle->post($url, [
    'form_params' => [
        'client_id' => $clientId,
        'client_secret' => $clientSecret,
        'resource' => 'https://graph.microsoft.com/',
        'grant_type' => 'client_credentials',
    ],
])->getBody()->getContents());

$accessToken = $token->access_token;
// Just for development test: Output Access-Token to check wether it is successfully generated (if not, credential data is not correct)
echo "AccessToken:".$accessToken;

$graph = new Graph();
$graph->setAccessToken($accessToken);

$user = $graph->createRequest("GET", "/users/SOME EMAIL ADDRESS")
->setReturnType(Model\User::class)
->execute();
  
echo "Hello, my name is {$user->getGivenName()}.";
```

This works for me perfectly

**Comment by SomeDevWeb** [comment]
**Thank a lot, basementmedia2!**

I tried it and I'm afraid I get the same result. I get what seems like a correct token, then NOTHING.
It could be something I am setting wrongly in the permission maybe.
O maybe I'm not calling the graph correctly or not including it the right way in the script... 

This is how I have the include. How do you use it?
`require 'vendor/msgraph-sdk-php/vendor/autoload.php';`

Can you please add this to your code to see what's outputted:
```
$graph = new Graph();
$graph->setAccessToken($accessToken);
**var_dump($graph);**
...
```

Mine writes this:

```
object(Microsoft\Graph\Graph)#29 (5) {
  ["_accessToken":"Microsoft\Graph\Graph":private]=>
  string(1375) "eyJ0eXAiOiJKV1QiLCJub25jZSI6InpPOUx6clJSZU.........."
  ["_apiVersion":"Microsoft\Graph\Graph":private]=>
  string(4) "v1.0"
  ["_baseUrl":"Microsoft\Graph\Graph":private]=>
  string(28) "https://graph.microsoft.com/"
  ["_proxyPort":"Microsoft\Graph\Graph":private]=>
  NULL
  ["_proxyVerifySSL":"Microsoft\Graph\Graph":private]=>
  NULL
}
```

**Thank you very much for the help!**

**Comment by basementmedia2** [comment]
Hi,

output of var_dump($accesstoken) seems to be ok.
Mine:

object(Microsoft\Graph\Graph)#29 (5) {
 ["_accessToken":"Microsoft\Graph\Graph":private]=> string(1626) "oiYzg1M2NhM2MtNDJiMy00M2E4LTlmN2YtMWExZmQ1Y2JjNDI3IiwidGVuYW50X3JlZ2lvbl9zY29wZSI6IkVVIiwidGlkIjoiM2FhMWJiYjMtN2QwZC00N2Y0LThiMjQtMGM5Yjk4YzAzZGEwIiwidXRpIjoibHB0U1k4X0FWMEtrSkhwSHRkTk1BQSIsInZlciI6IjEuMCIsIndpZHMiOlsiMDk5N2ExZDAtMGQxZC00YWNiLWI0MDgtZDVjYTczMTIxZTkwIl0sInhtc190Y2R0IjoxNTIyMDc1MzE5fQ.MKAYeuq1sn2YN-u9UBXnMDl_NGLdf8AsbtJK5DIyGm1VdsKkXPN17IwWsFbmJTQD7AdKdv8I06P_qCBBB585V315Rfhfa06CMVItOmgNDNkMpfP3q8HoQ_2mAveVeZp5LdqiWnhHGKvczfDaJBeIdZbf21VWrqABh5XeA-Vgu83zI6zkVS5hVj04Bk1W3FKwQ1BbVq4P5bu9xXdcvViLpunvKLEJUTZrRFhbbqW4BYF2cR3Rt3jCbyAU2h-8FIvEosZkr8i5mik0bKMUsoOEf37oJEPJcCTeDBjYHW77djo2533hRx3F8XtvkwLq5m2S8dAZY3R_zPtYNTrtibrchA" ["_apiVersion":"Microsoft\Graph\Graph":private]=> string(4) "v1.0" ["_baseUrl":"Microsoft\Graph\Graph":private]=> string(28) "https://graph.microsoft.com/" ["_proxyPort":"Microsoft\Graph\Graph":private]=> NULL ["_proxyVerifySSL":"Microsoft\Graph\Graph":private]=> NULL } 

I include the autoload this way:

require_once __DIR__ . '/vendor/autoload.php';

For me it looks that you have not set the correct permissions.

Log in to https://portal.azure.com/
Register your app (you have already)
Click on the app
choose API-Permissions (don't know the correct translation, in german "API-Berechtigungen"

For the example code i have set the following permissions
User.Read (delegiert)
User.Read.All(Application)

Don't know if you need both.
Our system admin had to set these permissions

**Comment by Ndiritu** [comment]
> So, I've asked an admin to grant all permission and I use: **User.ReadWrite.All | Application** | Read and write all users' full profiles | Yes | Granted for... **User.Read.All | Application** | Read all users' full profiles | Yes |  Granted for...
> 
> What other permission do I need please or what am I doing wrong?
> 
> I always get a valid token, but always a blank screen after, no other response, not even an error.
> 
> Here is my code:
> 
> use Microsoft\Graph\Graph; use Beta\Microsoft\Graph\Model as BetaModel;
> 
> $graph = new Graph();
> 
> $graph->setAccessToken($ms_accessToken);
> 
> $user = $graph->setApiVersion("beta") ->createRequest("GET", "/users/[my_user_email]") ->setReturnType(BetaModel\User::class) ->execute();
> 
> echo "Hello, I am $user->getGivenName() ";
> 
> Thank you!

@SomeDevWeb an exception should be thrown by the SDK if the request failed. Trying to figure out why that doesn't seem to be the case for you.

**Comment by basementmedia2** [comment]
Try this

$user = $graph->createRequest("GET", "/users/[your_email]")
->setReturnType(Model\User::class)
->execute();
print_r($user);
echo "Hello, my name is {$user->getGivenName()}.";

**Comment by basementmedia2** [comment]
By the way:
You can try/test requests also via the graph explorer:

https://developer.microsoft.com/en-us/graph/graph-explorer

When you press the "Modify Permissions"-button you see which permissions are required and can set them directly i think.

![grafik](https://user-images.githubusercontent.com/23479553/152750874-9f21cb47-b6be-42cf-934f-0204a12513b7.png)

**Comment by SomeDevWeb** [comment]
**Thanks all for your help and suggestions!**

I finally cracked it and will detail here maybe it helps others... ;)
So @basementmedia2, about graph explorer, I already tried it several times and that actually always returns the data correctly, which is a good test in itself that the system partly works, but that didn't help in sorting out the code side. 

Then @Ndiritu, that's also exactly what I was surprised about, why doesn't the graph sdk return any error or warning. NOTHING!
So, what I did was download again the whole file structure from github and copied them over the initial files I installed, cause I thought something was not installed correctly with composer or some files might be missing. 
I tried it again, and the same nothingness.... 

Then at some point it crossed my mind to just randomly check again the app credentials in the azure portal. 
Very surprisingly, there were completeley different clientid and tenantid. Obviously, I generated a new secretid and changed them all in code and indeed it finally worked after a week of research in vain. 

So, apologies if I'm wrong, but basically, there you go... No errors or warnings for something that is very obviously wrong and I think can be detected on the serve side and an error thown. More than that, I have no idea why those values changed and on top of that, obviously there was absolutely no notification to neither me or any admin about those changes. 

As I was saying, unfortunately I am not changing my mind at all about how much I DON'T want to have anything to do with MS.

Either way, thanks again for your help!!! Much appreciated.
